src/main/java/com/kryptnostic/datastore/services/EdmService.java

```diff
@@ -838,4 +854,182 @@ public AssociationDetails getAssociationDetails( UUID associationTypeId ) {
+            }
+        } );
+        edm.getAssociationTypes().forEach( at -> {
+            AssociationType existing = null;
+            EntityType et = at.getAssociationEntityType();
+            if ( et.wasIdPresent() )
+                existing = getAssociationType( et.getId() );
+            else {
+                UUID id = getTypeAclKey( et.getType() );
+                if ( id != null ) existing = getAssociationType( id );
+            }
+            if ( existing == null ) {
+                createEntityType( et );
+                createAssociationType( at, getTypeAclKey( et.getType() ) );
+            } else {
+                updateEntityType( et, existing.getAssociationEntityType() );
+                if ( !existing.getSrc().equals( at.getSrc() ) )
+                    addSrcEntityTypesToAssociationType( existing.getAssociationEntityType().getId(), at.getSrc() );
+                if ( !existing.getDst().equals( at.getDst() ) )
+                    addDstEntityTypesToAssociationType( existing.getAssociationEntityType().getId(), at.getDst() );
+            }
+        } );
+        edm.getSchemas().forEach( schema -> {
+            if ( schemaManager.checkSchemaExists( schema.getFqn() ) ) {
+                Schema existing = schemaManager.getSchema( schema.getFqn().getNamespace(), schema.getFqn().getName() );
+                if ( !existing.getEntityTypes().equals( schema.getEntityTypes() ) ) {
+                    schemaManager.addEntityTypesToSchema( schema.getEntityTypes().stream().map( et -> {
+                        return et.getId();
+                    } ).collect( Collectors.toSet() ), schema.getFqn() );
+                }
+                if ( !existing.getPropertyTypes().equals( schema.getPropertyTypes() ) ) {
+                    schemaManager.addPropertyTypesToSchema( schema.getPropertyTypes().stream().map( pt -> {
+                        return pt.getId();
+                    } ).collect( Collectors.toSet() ), schema.getFqn() );
+                }
+            } else {
+            schemaManager.createOrUpdateSchemas( schema );
+             }
+        } );
+    }
+
+    @Override
+    public EntityDataModel getEntityDataModelDiff( EntityDataModel edm ) {
+        Set<PropertyType> updatedPropertyTypes = Sets.newHashSet();
+        Set<EntityType> updatedEntityTypes = Sets.newHashSet();
+        Set<AssociationType> updatedAssociationTypes = Sets.newHashSet();
+        Set<Schema> updatedSchemas = Sets.newHashSet();
+        edm.getPropertyTypes().forEach( pt -> {
+            PropertyType existing = null;
+            if ( pt.wasIdPresent() ) {
+                if ( checkPropertyTypeExists( pt.getId() ) ) existing = getPropertyType( pt.getId() );
+            } else {
+                UUID id = getTypeAclKey( pt.getType() );
+                if ( id != null ) existing = getPropertyType( id );
+            }
+            if ( existing == null
+                    || !pt.getType().equals( existing.getType() )
```

Comment:
I'm skeptical about this type of equality check being needed elsewhere as this highly specific to our rules for supported diffs, but I think the point about moving the diff logic into some other location is valid as it will dramatically improve readability of this function.